ssl/ssl_lib.c

```diff
@@ -3889,6 +3891,103 @@ void SSL_set_not_resumable_session_callback(SSL *ssl,
                       (void (*)(void))cb);
 }
 
+static size_t ssl_record_padding_block_cb(SSL *s, int type, size_t len, void *arg)
+{
+    size_t block_size = (size_t)arg;
+    size_t remainder = len % block_size;
+
+    /* don't want to add a block of padding if we don't have to */
+    if (remainder == 0)
+        return 0;
+
+    return block_size - remainder;
+}
+
+/* optimized for a power of 2 */
+static size_t ssl_record_padding_block_cb2(SSL *s, int type, size_t len, void *arg)
+{
+    size_t mask = (size_t)arg;
+    size_t remainder = len & mask;
+
+    /* don't want to add a block of padding if we don't have to */
+    if (remainder == 0)
+        return 0;
+
+    return mask + 1 - remainder;
+}
+
+void SSL_CTX_set_record_padding_callback(SSL_CTX *ctx,
+                                         size_t (*cb) (SSL *ssl, int type,
+                                                       size_t len, void *arg))
+{
+    ctx->record_padding_cb = cb;
+}
+
+void SSL_CTX_set_record_padding_callback_arg(SSL_CTX *ctx, void *arg)
+{
+    ctx->record_padding_arg = arg;
+}
+
+void *SSL_CTX_get_record_padding_callback_arg(SSL_CTX *ctx)
+{
+    return ctx->record_padding_arg;
+}
+
+void SSL_CTX_set_block_padding(SSL_CTX *ctx, size_t block_size)
+{
+    if (block_size == 1 || block_size == 2) {
+        /* block size of 0 or 1 is basically no padding */
+        ctx->record_padding_cb = NULL;
+        ctx->record_padding_arg = 0;
+        return;
+    }
+    if ((block_size & (block_size - 1)) == 0) {
+        /* power of two optimized */
+        ctx->record_padding_cb = ssl_record_padding_block_cb2;
+        ctx->record_padding_arg = (void*)(block_size-1);
+        return;
+    }
+    /* some other value */
+    ctx->record_padding_cb = ssl_record_padding_block_cb;
+    ctx->record_padding_arg = (void*)block_size;
```

Comment:
I assume this is what @t-j-h is talking about...  I'm not sure I see this as an issue, as the important thing is that the callback knows what the void callback arg actually is.  I.e., as long as the callback function and the callback arg match, we're fine.

It *is* surprising that the user's callback and corresponding arg get tossed like this, but I suppose that it would be equally absurd to support this kind of block padding and the usual record padding callback and arg simultaneously.